circ.ks

```diff
@@ -1,35 +1,59 @@
 /////////////////////////////////////////////////////////////////////////////
 
 run once lib_ui.
 run once lib_util.
 
-IF Career():CANMAKENODES {
-  if obt:transition = "ESCAPE" or eta:periapsis < eta:apoapsis {
-    run node_apo(obt:periapsis).
-  } else {
-    run node_peri(obt:apoapsis).
-  }
-
-  run node.
+if Career():canMakeNodes and periapsis > max(body:atm:height,1000)
+{
+	utilRemoveNodes().
+	if obt:transition = "ESCAPE" or eta:periapsis < eta:apoapsis
+    	run node({run node_apo(obt:periapsis).}).
+  	else run node({run node_peri(obt:apoapsis).}).
+	uiBanner("Circ", "Circularized; e=" + round(ship:obt:eccentricity, 3)).
+}
+else if apoapsis > 0 and eta:apoapsis < eta:periapsis
+{
+	run once lib_staging.
+	run once lib_warp.
 
-  uiBanner("Circ", "Circularized; e=" + round(ship:obt:eccentricity, 3)).
-} ELSE {
-  sas off.
-  lock steering to ship:prograde.
-  uiBanner("Circ", "Coast to apoapsis.").
-  wait until utilIsShipFacing(ship:prograde:forevector).
-  run warp(eta:apoapsis - 30).
-  wait until utilIsShipFacing(ship:prograde:forevector).
-  // TODO be less lazy; calculate ideal delay for perfect circularization
-  wait until eta:apoapsis < 15.
-  uiBanner("Circ", "Burn to raise periapsis.").
-  set ship:control:pilotmainthrottle to max(1, 2 * ship:obt:eccentricity).
-  local height is ship:obt:apoapsis + 500.
-  WAIT UNTIL ship:obt:eccentricity < 0.01 or ship:obt:apoapsis > height.
-  set ship:control:pilotmainthrottle to 0.
-  sas on.
+	sas off.
+	set v0 to velocityAt(ship,time:seconds+eta:apoapsis):orbit.
+	lock steering to v0.
+	stagingPrepare().
+	// deltaV = required orbital speed minus predicted speed
+	set dv to sqrt(body:mu/(body:radius+apoapsis))-v0:mag.
+	set dt to burnTimeForDv(dv)/2.
+	uiBanner("Circ", "Coast to apoapsis.").
+	wait until utilIsShipFacing(v0).
+	warpSeconds(eta:apoapsis - dt - 30).
+	lock steering to prograde.
+	wait until utilIsShipFacing(prograde:forevector).
+	warpSeconds(eta:apoapsis - dt - 5).
+	wait until eta:apoapsis <= dt + 0.1.
+	uiBanner("Circ", "Burn to raise periapsis.").
+	lock throttle to (sqrt(body:mu/(body:radius+apoapsis))-ship:velocity:orbit:mag)*ship:mass/max(1,availableThrust).
+	local function circSteering {
+		if eta:apoapsis < eta:periapsis return prograde.
+		// pitch up a bit when we passed apopapsis to compensate for potentionally low TWR as this is often used after launch script
+		// note that ship's pitch is actually yaw in world perspective (pitch = normal, yaw = radial-out)
+		return prograde:vector+r(0,min(30,max(0,orbit:period-eta:apoapsis)),0).
+	}
+	lock steering to circSteering().
+	local maxHeight is ship:obt:apoapsis*1.01.
+	until ship:obt:eccentricity < 0.0005	// circular
+		or ship:obt:apoapsis > maxHeight	// something went wrong?
+		or eta:apoapsis > orbit:period/3 and eta:apoapsis < orbit:period*2/3 // happens with good accuracy
+	{
+		stagingCheck().
+		wait 0.5.
+	}
+	set ship:control:pilotmainthrottle to 0.
+	unlock all.
+	sas on.
```

Comment:
In some other places we check the original state for RCS and SAS.  Could we do that here?  I hadn't had this bite me because my vessels always answered true to the original if. :-)

Reply:
I vote for restoring RCS state, but always turn SAS on at the end of the script to stabilize the rocket - I hope that it chooses killrot.

Reply:
Restoring state minimizes side effects, which I think is a desirable quality.

As an example of why this is important, try this: `sas off. run circ_alt(100000). lock steering to retrograde.` :-)

Reply:
I disagree, but will follow established practices.